nixos/modules/services/x11/display-managers/gdm.nix

```
@@ -321,21 +322,31 @@ in
         session   include       login
       '';
 
+      # This would block password prompt when included by gdm-password.
+      # GDM will instead run gdm-fingerprint in parallel.
       login.fprintAuth = mkIf config.services.fprintd.enable false;
+
       gdm-fingerprint.text = mkIf config.services.fprintd.enable ''
         auth       required                    pam_shells.so
         auth       requisite                   pam_nologin.so
         auth       requisite                   pam_faillock.so      preauth
         auth       required                    ${pkgs.fprintd}/lib/security/pam_fprintd.so
-        auth       optional                    pam_permit.so
         auth       required                    pam_env.so
         auth       [success=ok default=1]      ${pkgs.gnome.gdm}/lib/security/pam_gdm.so
+        ${lib.optionalString
+          pamCfg.gdm-fingerprint.enableGnomeKeyring
+          "auth       optional                    ${pkgs.gnome-keyring}/lib/security/pam_gnome_keyring.so"}
+        # So that pam_gdm has something to skip over in the default branch when gnome-keyring is disabled.
+        auth       optional                    pam_permit.so
```

```suggestion
        ${if pamCfg.gdm-fingerprint.enableGnomeKeyring then
          "auth       optional                    ${pkgs.gnome-keyring}/lib/security/pam_gnome_keyring.so" else
          # So that pam_gdm has something to skip over in the default branch when gnome-keyring is disabled.
          "auth       optional                    pam_permit.so"}
```
So that the `pam_permit.so` line is only included when it needs to be.

---

Hmm, I am split on whether that is easier to read. But it gave me an idea – I wonder if it would make sense to remove the `pam_permit.so` altogether and use something like the following control for `pam_gdm.so`:

```nix
${if pamCfg.gdm-fingerprint.enableGnomeKeyring then "[success=ok default=1]" else "optional"}
```

---

I agree, that's preferable.

---

Decided to make both lines conditional instead, since that will be even cleaner. There are no further `auth` rules that could benefit from `pam_gdm.so` doing something.